src/service/viewer-impl.js

```diff
@@ -53,6 +53,14 @@ const VIEWER_ORIGIN_TIMEOUT_ = 1000;
 const TRIM_ORIGIN_PATTERN_ =
   /^(https?:\/\/)((www[0-9]*|web|ftp|wap|home|mobile|amp|m)\.)+/i;
 
+/**
+ * Pattern for removing any old "ampshare" parameters from a fragment.
+ * @const
+ * @private {!RegExp}
+ */
+const AMPSHARE_PATTERN =
+  /(^|&)ampshare[^&]*/g;
```

Comment:
I think we can avoid a new regex since we already parse the hash params (line 228). Instead, let's just store those in an ivar and check them later. Then you can reconstruct the fragment with the new `ampshare` param.

```js
this.hashParams_ = {};

...

if (this.win.location.hash) {
  parseParams_(this.win.location.hash, this.hashParams_);
  // this.params_ should include all params, including hash params.
  Object.assign(this.params_, this.hashParams_); 
}

...

this.hashParams_['ampshare'] = canonicalUrl;
this.win.history.replaceState({}, '', '#' + serializeQueryString(this.hashParams_));
```